### ISSUE TITLE
Properly classify infra dependencies that show up as "at least one terraformer pod still exists" errors

### ISSUE THREAD

**Issue by stoyanr** [gardener/gardener#4838]
**How to categorize this issue?**

/area networking ops-productivity
/kind enhancement
/priority 3

**What would you like to be added**:

When deleting a shoot that has infrastructure dependencies, we often see the following error:
```
task "Waiting until shoot infrastructure has been destroyed" failed: Failed to delete Infrastructure shoot--hc-dev--i30352895-hxtp/i30352895-hxtp: Error deleting infrastructure: retry failed with context deadline exceeded, last error: at least one terraformer pod still exists: i30352895-hxtp.infra.tf-destroy-vlg4g
```

In the logs of the terraformer pod we see something similar to (e.g. on AWS):
```
aws_security_group.nodes: Still destroying... [id=sg-0730b480742605e3e, 20s elapsed]
aws_subnet.nodes_z0: Still destroying... [id=subnet-0f14a6bef28063422, 20s elapsed]
aws_security_group.nodes: Still destroying... [id=sg-0730b480742605e3e, 30s elapsed]
aws_subnet.nodes_z0: Still destroying... [id=subnet-0f14a6bef28063422, 30s elapsed]
aws_security_group.nodes: Still destroying... [id=sg-0730b480742605e3e, 40s elapsed]
aws_subnet.nodes_z0: Still destroying... [id=subnet-0f14a6bef28063422, 40s elapsed]
```

Further investigation usually reveals that there are infrastructure dependencies that would need to be manually cleaned up, e.g. EFS mount targets on AWS.

```
gardenctl aws ec2 describe-network-interfaces --filters Name=group-id,Values=sg-0730b480742605e3e
...
NETWORKINTERFACES       eu-central-1a   EFS mount target for fs-7365ef28 (fsmt-68709b30)   
...
```

Unfortunately, due to the generic nature of the error message "at least one terraformer pod still exists", it is impossible to classify such errors as infra dependencies (`ERR_INFRA_DEPENDENCIES`) based on only on the error message.

We should properly determine the root cause and classify such issues as `ERR_INFRA_DEPENDENCIES`, perhaps by looking for known infrastructure dependencies (e.g. EFS mount targets on AWS) when such terraform errors occur.

**Why is this needed**:
Such errors occur quite often in our production landscapes and since they are not properly classified they clutter the dashboard and require effort from our operators to inform the customer and get rid of them.

**Comment by ialidzhikov** [comment]
Similar to https://github.com/gardener/gardener-extension-provider-aws/issues/211

**Comment by stoyanr** [comment]
The current idea is to check if the terraform logs contains messages similar to:

```
aws_security_group.nodes: Still destroying... [id=sg-0730b480742605e3e, 40s elapsed]
aws_subnet.nodes_z0: Still destroying... [id=subnet-0f14a6bef28063422, 40s elapsed]
```

and report an error of type `ERR_INFRA_DEPENDENCIES` if the `Still destroying...` message is:
* About a subnet or a security group.
* A sufficiently long time had elapsed (e.g. longer than the configured deletion timeout of 5m for subnets).

**Comment by stoyanr** [comment]
To reproduce this, it's easier to create an instance for the subnet rather than an "EFS mount target", e.g.:

* Create a normal AWS shoot with e.g. a single node
* Note the security group id and the `nodes` subnet id in the status of the `infrastructure` resource in the seed
* Get the AWS existing instance(s) by filtering on the subnet id:
  ```
  aws ec2 describe-instances --filters "Name=subnet-id,Values=subnet-0893f73bce3a14eea"
  ```
  Note the image id, instance type, and key name of this instance

* Create a new instance in the same security group id and subnet id, and with the same parameters as the existing one:
  ```
  aws ec2 run-instances --image-id ami-06672bb04fdda3653 --count 1 --instance-type m5.large --key-name shoot--dev--i500152-aws-ssh-publickey --security-group-ids sg-0a6cd54f5a3d7c0e3 --subnet-id subnet-0893f73bce3a14eea
  ```
  Note the instance id of this instance.

* Try to delete the shoot. The first deletion should fail with an error that is appropriately classified as `ERR_INFRA_DEPENDENCIES`. This is because during the first terraform pod execution, the custom timeout of 5m is respected. The shoot will be in `Delete Failed`. The infrastructure controller may still try to run the terraformer pod for a second time before the shoot status has been updated in the cluster resource.
* Retry the shoot deletion by setting `gardener.cloud/operation=retry`. This will cause a second reconciliation that will be stuck in `Delete Processing` for some time with the above error. This is because during subsequent terraformed pod executions, the custom timeout of 5m is not respected but instead the default timeout of 20m applies. However, the infrastructure deletion itself times out after 15m, before the terraformer pod has managed to time out, and the terraformer logs still don't contain any errors, only `Still destroying...` messages. This state will likely persist until you manually delete the extra instance with:
  ```
  aws ec2 terminate-instances --instance-ids i-0b8c002d6b5fd62ee
  ```

**Comment by stoyanr** [comment]
To implement the above suggestion, either the existing `findTerraformerErrors` function (https://github.com/gardener/gardener/blob/8c603101963bf67b90dd38b69fd3add96e2a911c/extensions/pkg/terraformer/errors.go#L25) could be enhanced or a new similar function could be added that would try to parse terraformer logs in order to find `Still destroying...` messages that match the above conditions.

**Comment by stoyanr** [comment]
I added some more log messages and double checked that in this scenario, the logs are indeed retrieved and processed, but since `findTerraformErrors` doesn't find any error messages in the logs, no classifiable errors are reported.

**Comment by stoyanr** [comment]
In a team discussion we agreed that since the issue is fixed in `terraform` 0.15 (see hashicorp/terraform/issues/27452) it's best to upgrade to this version. Unfortunately it's not possible to do this in a single move, due to the terraform upgrade policy we should go through all intermediate versions (0.13 and 0.14) before we could upgrade to 0.15.

I already started this process with gardener/terraformer/pull/105 which upgrades terraformer to 0.13.7.

**Comment by stoyanr** [comment]
I tested with an AWS provider extension image from gardener/gardener-extension-provider-aws/pull/443 and I am no longer able to reproduce the issue by following the above steps. I see that the timeout of 5m is appropriately honored by `terraform` not only during the initial execution, but also during each subsequent execution. This confirms that upgrading terraform to 0.15 effectively fixes this issue.

**Comment by stoyanr** [comment]
I opened 3 different PRs for upgarding terraform to 0.13, 0.14, and 0.15 respectively:

* gardener/terraformer/pull/105
* gardener/terraformer/pull/106
* gardener/terraformer/pull/107

As can be seen from the PRs, I tested all 5 provider extensions with each one of them successfully, including updating from a state created with 0.12.31. I had to do some changes / adaptations for 0.13, but for 0.14 and 0.15 there were none or very minor changes. Also the upgrade guides to 0.14 and 0.15 suggest very few incompatible changes (see https://www.terraform.io/upgrade-guides/0-14.html and https://www.terraform.io/upgrade-guides/0-15.html).

Because of this, it might be a good idea to upgrade to terraform 0.15 directly, instead of going through the intermediate versions 0.13 and 0.14, as this would save quite some time and effort.

/cc @ialidzhikov

**Comment by stoyanr** [comment]
Closed by mistake, reopening.

**Comment by ialidzhikov** [comment]
This issue should be resolved with the adoption of `terraform@0.15.5` in the provider extensions (especially for provider-aws it is done with https://github.com/gardener/gardener-extension-provider-aws/pull/505). Hence, we can close this issue now.

/close
